V8 v8.5 release blog post (#445)
v8/v8.dev: src/blog/v8-release-85.md

---
title: 'V8 release v8.5'

> author: 'Sathya Gunasekaran'

Reviewer: Change to your name

Reviewer 2: And add a funny/pun tagline like in other posts please.

> avatars:

Reviewer 2: Add your avatar to `img` folder in 2 sizes and reference here.

Reviewer 3: @zeynepCankara Only the images in `src/_img/avatars/*` are needed; please remove the ones you added in `src/_img/*.jpg`. Also, could you check the dimensions of the avatar images (it doesn't look like they're square)? They should match the other avatars.

date: 2020-07-21
tags:
 - release

> description: 'V8 release v8.5'

Reviewer 2: This should include a short description of features, see other posts for examples.

Author: Changed it

tweet:
---
Every six weeks, we create a new branch of V8 as part of our [release process](https://v8.dev/docs/release-process). Each version is branched from V8’s Git master immediately before a Chrome Beta milestone. Today we’re pleased to announce our newest branch, [V8 version 8.5](https://chromium.googlesource.com/v8/v8.git/+log/branch-heads/8.5), which is in beta until its release in coordination with Chrome 85 Stable in several weeks. V8 v8.5 is filled with all sorts of developer-facing goodies. This post provides a preview of some of the highlights in anticipation of the release.

> JavaScript

Reviewer 2: These should be formatted as headings.

```javascript
instance.exports.swap(1, 2) // Returns [2, 1].
```

Conversely, if a JavaScript function returns an array (or any iterator), it can be imported and called as a multi-return function inside the WebAssembly module:

```javascript
new WebAssembly.Instance(module, { imports: { swap: (x, y) => [y, x] } });
```

```wasm
(func $main (result i32 i32)
  i32.const 0
  i32.const 1
  call $swap
)
```

More importantly, toolchains can now use this feature to generate more compact and faster code within a WebAssembly module.

Support for JS BigInts
WebAssembly support for [converting WebAssembly I64 values from and to JavaScript BigInts](https://github.com/WebAssembly/JS-BigInt-integration) has been shipped and is available for general use as per the latest change in the official standard.

Thereby WebAssembly functions with I64 parameters and return values can be called from JavaScript without precision loss:

```wasm
(module
  (func $add (param $x i64) (param $y i64) (result i64)
    local.get $x
    local.get $y
    i64.add)
  (export "add" (func $add)))
```

From JavaScript, only BigInts can be passed as I64 parameter:

```javascript

> WebAssembly.instantiateStreaming(fetch(‘i64.wasm’))

Reviewer 2: ```suggestion
WebAssembly.instantiateStreaming(fetch('i64.wasm'))
```

```javascript
new WebAssembly.Instance(module, { imports: { swap: (x, y) => [y, x] } });
```

```wasm
(func $main (result i32 i32)
  i32.const 0
  i32.const 1
  call $swap
)
```

More importantly, toolchains can now use this feature to generate more compact and faster code within a WebAssembly module.

### Support for JS BigInts

WebAssembly support for [converting WebAssembly I64 values from and to JavaScript BigInts](https://github.com/WebAssembly/JS-BigInt-integration) has been shipped and is available for general use as per the latest change in the official standard.

Thereby WebAssembly functions with I64 parameters and return values can be called from JavaScript without precision loss:

```wasm
(module
  (func $add (param $x i64) (param $y i64) (result i64)
    local.get $x
    local.get $y
    i64.add)
  (export "add" (func $add)))
```

From JavaScript, only BigInts can be passed as I64 parameter:

```javascript
WebAssembly.instantiateStreaming(fetch('i64.wasm'))
  .then(({ module, instance })) => {
    instance.exports.add(12n, 30n); // result = 42n;
    instance.exports.add(12, 30); // throws TypeError, parameters are not of type BigInt.
  });

> ```

Reviewer 3: The V8 API section is missing (see other release notes)

```javascript
x &&= y;
// Roughly equivalent to x && (x = y)
x ||= y;
// Roughly equivalent to x || (x = y)
x ??= y;
// Roughly equivalent to x ?? (x = y)
```

Note that, unlike mathematical and bitwise compound assignment operators, logical assignment operators only conditionally perform the assignment. Please read our explainer for a more in-depth explanation.

## WebAssembly

### Liftoff shipped on all platforms

Since v6.9, [Liftoff](https://v8.dev/blog/liftoff) has been used as the baseline compiler for WebAssembly on Intel platforms (Chrome enabled it on desktop systems since M-69). Since we were concerned about memory increase (because of more code being generated by the baseline compiler), we held it back for mobile systems so far. After some experimentation in the last months, we are confident that the memory increase is negligible for most cases, hence we finally enable Liftoff by default on all architectures, bringing increased compilation speed, especially on arm devices (32- and 64-bit). Chrome follows along and ships Liftoff in M-85.

> Multi-value support shipped

Reviewer 3: Should be a header. Are you testing + previewing this locally? Doing so might make it easier to spot these kinds of things

---
title: 'V8 release v8.5'
author: 'Zeynep Cankara, tracking some Maps'
avatars:
 - 'zeynep-cankara'
date: 2020-07-21
tags:
 - release

> description: 'V8 release v8.5 features Promise.any, AggregateError, String.prototype.replaceAll, logical assignment operators, Liftoff shipped on all platforms, multi-value support shipped and support for JS BigInts'

Reviewer 2: Let's reword to shorten this up a bit.

Reviewer 2: Maybe at least

```suggestion
description: 'V8 release v8.5 features Promise.any, .replaceAll on strings, logical assignment operators, WebAssembly multi-value and BigInt support, and performance improvements'
```

Reviewer 3: Could also use the `String#replaceAll` notation (we're doing it all over the place on v8.dev)

Reviewer 2: Oh yeah right, that's probably even better.

Reviewer 2: Updated.

Reviewer 2: (this got accidentally reverted in a follow-up commit, marking as unresolved)

## JavaScript

### `Promise.any` and `AggregateError`

`Promise.any` is a promise combinator that resolves the resulting promise as soon as one of the input promises is fulfilled.

```js
const promises = [
  fetch('/endpoint-a').then(() => 'a'),
  fetch('/endpoint-b').then(() => 'b'),
  fetch('/endpoint-c').then(() => 'c'),
];
try {
  const first = await Promise.any(promises);
  // Any of the promises was fulfilled.
  console.log(first);
  // → e.g. 'b'
} catch (error) {
  // All of the promises were rejected.
  console.assert(error instanceof AggregateError);
  // Log the rejection values:
  console.log(error.errors);
}
```

If all input promises are rejected, the resulting promise is rejected with an `AggregateError` object containing an `errors` property which holds an array of rejection values.

> Please see our explainer for more.

Reviewer 3: Missing link?

### `String.prototype.replaceAll`

`String.prototype.replaceAll` provides an easy way to replace all occurrences of a substring without creating a global `RegExp`.

```js
const queryString = 'q=query+string+parameters';

// Works, but requires escaping inside regular expressions.
queryString.replace(/\+/g, ' ');
// → 'q=query string parameters'

// Simpler!
queryString.replaceAll('+', ' ');
// → 'q=query string parameters'
```

> Please see our explainer for more.

Reviewer 3: Missing link?
